### PR TITLE
Twitter iOS 5 Integration

### PR DESCRIPTION
**Pull request by brianantonelli**
Plugin to allow you to hook into iOS 5's Twitter integration.

### CODE REVIEW

**Comment by BenJam** [comment]
like it.

**Comment by peterbenoit** [comment]
very nice!

**Comment by gingerbeardman** [comment]
@maxogden is there any way compose can take arguments so you can at least attach default text, image or url to the tweet? Thanks

**Comment by brianantonelli** [comment]
@gingerbeardman Use the sendTweet function, it lets you specify all of that.

**Comment by gingerbeardman** [comment]
@brianantonelli I wanted to prefill the compose view, not send. I've just committed my changes to add this. thanks

**Comment by gingerbeardman** [comment]
@brianantonelli Ah, I now see sendTweet doesn't actually send until the user hits the send button. The name made me imagine sending without any interaction. Thanks for the headsup.

**Comment by sromalewski** [comment]
This seems to be a great plugin, thanks for putting it together.  But when I try to use the sendTweet function, I keep getting the error "tweet failure: Image could not be added" no matter what I do for the arguments.  I've tried a valid image url, no url (i.e., just ""), etc but the error is always there.  THoughts?

**Comment by sromalewski** [comment]
Btw, I've encountered two other issues.

One is that when I analyze my app in Xcode (4.2), it displays two "potential memory leaks" with TwitterPlugin.m.  At lines 61 and 63 it says "Method returns an Objective-C object with a +1 retain count" and "Object leaked: object allocated and stored into 'img' is not referenced later in this execution path and has a retain count of +1".  And at lines 97 and 103 it says "Method returns an Objective-C object with a +1 retain count" and "Object leaked: object allocated and stored into 'tweetComposeViewController' is not referenced later in this execution path and has a retain count of +1".

The other issue is that I get a "fatal error" when I try to run my app on either a 4.2 or 4.3 simulator. (This is after I've implemented the fix to resolve the "EXC_BAD_ACCESS" problem.)  In my console, it says "dyld: Library not loaded: /System/Library/Frameworks/Accounts.framework/Accounts".  But I've added the Accounts framework, per the plugin instructions.

**Comment by sromalewski** [comment]
I changed Accounts.framework and Twitter.framework to "Optional" and the app loaded ok on the 4.0, 4.2, and 4.3 simulators.

The memory leak issues are still there, though.

**Comment by sromalewski** [comment]
I determined that the image error message was because the message I was sending programmatically via sendTweet was too long. If I make sure it's under 140 characters first, then it'll work. But that's too bad, because I'd like to give the user the option to pare it down as they like, rather than me truncating it first. Is there a way to allow the tweet box to open even if the pre-set tweet is more than 140 chars?

**Comment by brianantonelli** [comment]
@sromalewski I've got a pull request in that addresses this.

https://github.com/phonegap/phonegap-plugins/pull/327

**Comment by sromalewski** [comment]
Thanks!  I'll try this out.  Any thoughts on the memory leaks?

**Comment by gingerbeardman** [comment]
Are you using ARC (automatic reference counting) or not? I nuked a load of release statements in PhoneGap because I am using ARC and I am no longer getting any warnings.

**Comment by sromalewski** [comment]
Interesting.  I read somewhere that Phonegap doesn't use ARC so when I create a new project in Xcode 4.2 I explicitly uncheck that option.  I'll try it with the ARC option checked.

Is there an easy way to use ARC with an existing Xcode 4.2 project? That is, change a setting in a plist file, perhaps, so I don't need to go through the process of recreating the Xcode project?

**Comment by brianantonelli** [comment]
If you're using ARC you'll run into these errors, PG wasnt using ARC when I wrote this plugin.

**Comment by brianantonelli** [comment]
@sromalewski I wouldn't use ARC in a Phonegap project, especially if you're using other peoples plugins.

**Comment by sromalewski** [comment]
@brianantonelli Thanks for clarifying.  But I'm not using ARC and I'm still getting the errors.  Maybe I can ignore them?  (The app compiles fine and runs ok in the simulator and on an iPhone.  Nonetheless, they're worrisome.)

**Comment by brianantonelli** [comment]
@sromalewski  I see the leaks, I'm looking into it..

**Comment by sromalewski** [comment]
Sounds good, thanks!

**Comment by brianantonelli** [comment]
@sromalewski Pull request updated to include memory leak fix.

**Comment by sromalewski** [comment]
Thanks.  I've tried your fix (on the simulator) and the memory leaks are gone. Thanks!

But I'm not sure the test for tweet length is working. Or maybe I'm not implementing it right.  Here's my code:

function shareToTwitter() {
   var landmark_name, tweetTextNY, photo, imgtoTweet;
   tweetTextNY = landmark_name + ', a #NYC #landmark.';

```
if (photo === '50px-nopicture.jpg') {
    imgtoTweet = 'https://s3.amazonaws.com/LandmarksNYphotos/icon100x100_whitebackground_withlogo.png';
} else {
    imgtoTweet = 'https://s3.amazonaws.com/LandmarksNYphotos/' + photo;
}

window.plugins.twitter.isTwitterAvailable(function(r){
                                          if (r === 1){
                                          window.plugins.twitter.sendTweet(
                                                                           function(s){ console.log("*****tweet success: " + s); }, 
                                                                           function(e){ console.log("*****tweet failure: " + e); }, 
                                                                           tweetTextNY, 
                                                                           'http://spatialityapps.com',
                                                                           imgtoTweet);
                                          } else {
                                          // Do something else
                                          }
                                          });        
```

}

If the length of tweetTextNY is relatively short (say, less than 80 chars), the tweetsheet shows up regardless of how long the imgtoTweet url is.  But if the tweetTextNY length is longer, the error message I get is always "tweet failure: Image could not be added".  

I also tried just using a long string of letters for tweetTextNY (about 200 chars), and I left imgtoTweet = ''.  And I still get the "tweet failure: Image could not be added" error.  But if I include the long string for tweetTextNY and use a valid url for imgtoTweet, the tweetsheet opens fine, but there's no message in the sheet (the long string is ignored), and the image is attached.

In other words, the new TwitterPlugin.m file doesn't seem to do the test for the tweet text, it just checks for the image.

Actually, if I set tweetTextNY = ''; or I leave the URL (3rd argument) blank, but set the imgtoTweet to a valid URL, the tweetsheet opens ok.  But anytime I set the imgtoTweet = ''; I get the "tweet failure: Image could not be added" error.  Is there a way to not attach an image?

**Comment by brianantonelli** [comment]
If you aren't attaching an image don't pass an argument for it. Also, I assume you may get an error if your tweet text is close to 140 char and you attach a link which hit the character limit after twitter shortens it and appends it to the tweet

**Comment by sromalewski** [comment]
But, as my last paragraph above explains, if I don't pass an argument for it (whether it's blank, or missing, or null) I get the "Image could not be added" error.  So not passing an argument is not helping.

Also, as my examples point out, it doesn't seem like the fix for checking the length of the tweet is working.  I never get the "Tweet is too long" error regardless of how long I make the text for the body of the tweet.

**Comment by sromalewski** [comment]
Actually, the same thing happens with the demo code that comes with the plugin.  The following examples all return "Image could not be added":

tweet:function(){
    TwitterDemo.log("wait..");
    window.plugins.twitter.sendTweet(
                                     function(s){ TwitterDemo.log("tweet success"); }, 
                                     function(e){ TwitterDemo.log("tweet failure: " + e); }, 
                                     "Tweety Poo", 
                                     "https://github.com/brianantonelli",
                                     " ");
},

OR
   window.plugins.twitter.sendTweet(
                                     function(s){ TwitterDemo.log("tweet success"); }, 
                                     function(e){ TwitterDemo.log("tweet failure: " + e); }, 
                                     "Tweety Poo", 
                                     "https://github.com/brianantonelli",
                                     "");

OR
   window.plugins.twitter.sendTweet(
                                     function(s){ TwitterDemo.log("tweet success"); }, 
                                     function(e){ TwitterDemo.log("tweet failure: " + e); }, 
                                     "Tweety Poo", 
                                     "https://github.com/brianantonelli");

So, I'm not sure how not to pass an argument for it.  The TwitterPlugin.js seems to be designed to handle this (if(typeof imageAttach === "undefined") imageAttach = "";), but either I'm implementing it wrong in my app js file or it's not working right.

**Comment by sromalewski** [comment]
I've moved on from the tweet length and image issues -- I decided to do two things: 1) use an if statement to check myself if the tweet text was too long and replace it with someone short and generic, and 2) use the native addImage option in obj-c to attach the image from a local file in the application which avoids having to grab the image via a URL.

I have a separate question, though, regarding performance. It takes a moment or two for the tweetsheet to be displayed.  I'm not sure if that's because this is a plugin and not completely native, or if performance is slow even with purely native apps. But I've tried to add a "Loading..." notification until the tweetsheet is displayed but it doesn't work. I call navigator.notificationEx.loadingStart(); right at the beginning of my shareToTwitter() function, but it never gets displayed.  Maybe it needs to be called asynchronously? as in:

function shareToTwitter() {
loadingCall();
window.plugins.twitter.sendTweet(
                                                                           function(s){ console.log("****_tweet success: " + s); }, 
                                                                           function(e){ console.log("**_**tweet failure: " + e); }, 
                                                                           tweetTextNY, 
                                                                           'http://spatialityapps.com',
                                                                           imgtoTweet);
navigator.notificationEx.loadingStop();
}

function loadingCall(){
navigator.notificationEx.loadingStart();
}

**Comment by brianantonelli** [comment]
@sromalewski I've reworked how the tweet is composed and it's now working correctly. 

I've created several example compose functions: 
- Blank
- Text
- Text and image
- Text and URL
- Text, image and URL

Regarding the lag in the dialog appearing, that usually only occurs when you're attaching an image or URL as some additional processing has to be done.. The loading indicator won't work because the stop loading call won't fire until after the sheet is dismissed. I think this is just as designed for the way Twitter is built into iOS.

The pull has been updated to include these changes. https://github.com/phonegap/phonegap-plugins/pull/327/commits

**Comment by sromalewski** [comment]
Thanks very much for the follow up.

I tried your demo code, and I still get the "Image could not be added" error for the following two examples:
<li><a href="#" id="tweet1">tweet (text, img, url)</a></li>
<li><a href="#" id="tweet2">tweet (text, img)</a></li>

But if I use the approach I mentioned above (using an if statement to check length of the tweet text, and modifying the TwitterPlugin.m file to use a local image instead of a URL-based image), then it works fine.

Here's the modification to the m file:
    if(imageAttach != nil){
        // Note that the image is loaded syncronously
        // Steve Romalewski changed to accomodate local images instead of URL based
        //UIImage *img = [[UIImage alloc] initWithData:[NSData dataWithContentsOfURL:[NSURL URLWithString:imageAttach]]];

```
    //ok = [tweetViewController addImage:img];
    ok = [tweetViewController addImage:[UIImage imageNamed:imageAttach]];
    //[img release];
    if(!ok){
        errorMessage = @"Image could not be added";
    }
}
```

... then in my JavaScript file I set a variable for the imageAttach element that points to a file in my www folder:
     imgtoTweet = 'www/images/photos/' + photo;

To deal with the lag in displaying the tweetsheet, I put a "navigator.notification.alert('Loading tweet.');" line at the start of my Twitter function.  Even though the actual alert doesn't display until the tweetsheet displays, it has the effect of causing the app to just slightly dim (hopefully indicating to the user that something's happening) until the tweetsheet loads.  Here's a screenshot of the app dimmed just before the alert/tweetsheet fires: http://dl.dropbox.com/u/20314275/twitter_app_dimmed.PNG  And here's what it looks like when the alert is displayed: http://dl.dropbox.com/u/20314275/twitter_app_alert.PNG

Btw, I've noticed that when I use the plugin to add an image to the tweetsheet, the image is always obscured under the paperclip on the tweetsheet (some sort of default graphic image of a compass covers up my image).  With other apps that have integrated Twitter, the image is always displayed as-is under the paper clip.  Here are two examples.  With your plugin (image is obscured): http://dl.dropbox.com/u/20314275/twitterplugin_imageobscured.PNG  With another app that uses Twitter (image is ok): http://dl.dropbox.com/u/20314275/twitter_imagedisplayed.PNG  What do you think?

**Comment by brianantonelli** [comment]
@sromalewski I guess i missed the part where you were trying to load a local image, this code obviously only supported attaching remote images. I'll try to add support for remote as well as local image attachments. 

Regarding the dimming to indicate loading, I'm still not sure what the best way is to handle this.. Let me noodle on it.

The image getting covered up by the compass is because a URL is being attached as well as the image. If you just attach an image it should be displayed instead of the compass (Safari icon). I attach the image after the URL; however, it looks like the SDK will always layer the URL attachment over the image attachment when they both exist..

**Comment by sromalewski** [comment]
Thanks.  Looks like you're right re: the image being obscured.  I omitted the URL and the image under the paper clip shows up fine.

The other benefit of omitting the URL is that it appears that the tweetsheet loads quicker (at least in the simulator; I'll test more on the device).  And I can still add the URL simply by including it in the tweetText variable, so it's the best of both worlds.

Nonetheless, I'm still going to use "navigator.notification.alert('Loading tweet.');" in my twitter function just in case.  If you have other workarounds I'd be glad to test them.

**Comment by gingerbeardman** [comment]
I assume that when you attach a URL there is some back and forth at the OS level with twitter as it shortens the URL?

**Comment by brianantonelli** [comment]
@sromalewski I've updated the pull request with code that differentiates between local and remote images.
@gingerbeardman Comparing the load time for the tweet sheet between remote and local images shows that the local image tweets load the sheet almost instantly. Twitter must be processing something with the image too.
